File: website/src/features/shipment/domain/entities/Shipment.ts
/**
 * Shipment Entity
 * Domain entity representing a shipment in the system
 * Contains business rules and validations for shipment data
 */

import type {
  ShipmentStatus,
  LinkedOrder,
  OneTimeSupplier,
} from '../types';

// ===== Core Shipment Entity =====

/** Base shipment properties shared across contexts */
export interface ShipmentBase {
  shipmentId: string;
  shipmentNumber: string;
  trackingNumber: string | null;
  shippedDate: Date;
  status: ShipmentStatus;
  notes: string | null;
  createdAt: Date;
  createdBy: string | null;
}

/** Supplier information for a shipment */
export interface ShipmentSupplier {
  supplierId: string | null;
  supplierName: string;
  supplierPhone: string | null;
  supplierEmail: string | null;
  supplierAddress: string | null;
  isRegisteredSupplier: boolean;
}

/** Shipment item entity */
export interface ShipmentItemEntity {
  itemId: string;
  productId: string;
  productName: string;
  sku: string;
  quantityShipped: number;
  unitCost: number;
  totalAmount: number;
}

/** Complete shipment entity with all details */
export interface Shipment extends ShipmentBase, ShipmentSupplier {
  totalAmount: number;
  items: ShipmentItemEntity[];
  hasOrders: boolean;
  orderCount: number;
  orders: LinkedOrder[];
  canCancel: boolean;
}

/** Shipment list item entity (summary for list view) */
export interface ShipmentListItemEntity extends ShipmentBase {
  supplierId: string | null;
  supplierName: string;
  itemCount: number;
  hasOrders: boolean;
  linkedOrderCount: number;
}

// ===== Create Shipment Entity =====

/** New shipment item for creation */
export interface NewShipmentItem {
  orderItemId: string;
  orderId: string;
  orderNumber: string;
  productId: string;
  productName: string;
  sku: string;
  quantity: number;
  maxQuantity: number;
  unitPrice: number;
}

/** New shipment entity for creation */
export interface NewShipment {
  items: NewShipmentItem[];
  trackingNumber?: string;
  notes?: string;
  // Either order-based or supplier-based
  orderId?: string;
  supplierId?: string;
  oneTimeSupplier?: OneTimeSupplier;
}

// ===== Entity Class Implementation =====

/**
 * ShipmentEntity class with business logic methods
 */
export class ShipmentEntity implements Shipment {
  shipmentId: string;
  shipmentNumber: string;
  trackingNumber: string | null;
  shippedDate: Date;
  status: ShipmentStatus;
  totalAmount: number;
  notes: string | null;
  createdBy: string | null;
  createdAt: Date;
  supplierId: string | null;
  supplierName: string;
  supplierPhone: string | null;
  supplierEmail: string | null;
  supplierAddress: string | null;
  isRegisteredSupplier: boolean;
  items: ShipmentItemEntity[];
  hasOrders: boolean;
  orderCount: number;
  orders: LinkedOrder[];
  canCancel: boolean;

  constructor(data: Shipment) {
    this.shipmentId = data.shipmentId;
    this.shipmentNumber = data.shipmentNumber;
    this.trackingNumber = data.trackingNumber;
    this.shippedDate = data.shippedDate;
    this.status = data.status;
    this.totalAmount = data.totalAmount;
    this.notes = data.notes;
    this.createdBy = data.createdBy;
    this.createdAt = data.createdAt;
    this.supplierId = data.supplierId;
    this.supplierName = data.supplierName;
    this.supplierPhone = data.supplierPhone;
    this.supplierEmail = data.supplierEmail;
    this.supplierAddress = data.supplierAddress;
    this.isRegisteredSupplier = data.isRegisteredSupplier;
    this.items = data.items;
    this.hasOrders = data.hasOrders;
    this.orderCount = data.orderCount;
    this.orders = data.orders;
    this.canCancel = data.canCancel;
  }

  // ===== Business Logic Methods =====

  /**
   * Check if shipment is editable
   */
  isEditable(): boolean {
    return this.status === 'pending' || this.status === 'process';
  }

  /**
   * Check if shipment can be cancelled
   */
  isCancellable(): boolean {
    return this.canCancel && this.status !== 'cancelled' && this.status !== 'complete';
  }

  /**
   * Get total quantity shipped
   */
  getTotalQuantity(): number {
    return this.items.reduce((sum, item) => sum + item.quantityShipped, 0);
  }

  /**
   * Check if shipment has supplier contact info
   */
  hasSupplierContact(): boolean {
    return !!(this.supplierPhone || this.supplierEmail);
  }

  /**
   * Get status display text
   */
  getStatusDisplay(): string {
    const statusMap: Record<ShipmentStatus, string> = {
      pending: 'Pending',
      process: 'In Process',
      complete: 'Complete',
      cancelled: 'Cancelled',
    };
    return statusMap[this.status] || this.status;
  }

  /**
   * Check if shipment is linked to orders
   */
  isOrderLinked(): boolean {
    return this.hasOrders && this.orderCount > 0;
  }

  /**
   * Calculate item subtotal
   */
  getItemSubtotal(itemId: string): number {
    const item = this.items.find((i) => i.itemId === itemId);
    return item ? item.quantityShipped * item.unitCost : 0;
  }

  /**
   * Recalculate total amount from items
   */
  recalculateTotalAmount(): number {
    return this.items.reduce((sum, item) => sum + item.totalAmount, 0);
  }
}

// ===== Factory Functions =====

/**
 * Create empty shipment item entity
 */
export function createEmptyShipmentItem(): NewShipmentItem {
  return {
    orderItemId: '',
    orderId: '',
    orderNumber: '-',
    productId: '',
    productName: '',
    sku: '',
    quantity: 1,
    maxQuantity: 0,
    unitPrice: 0,
  };
}

/**
 * Create new shipment item from product search
 */
export function createShipmentItemFromProduct(
  product: { product_id: string; product_name: string; sku: string; stock: { quantity_on_hand: number }; price: { cost: number } }
): NewShipmentItem {
  return {
    orderItemId: `search-${product.product_id}-${Date.now()}`,
    orderId: '',
    orderNumber: '-',
    productId: product.product_id,
    productName: product.product_name,
    sku: product.sku,
    quantity: 1,
    maxQuantity: product.stock.quantity_on_hand,
    unitPrice: product.price.cost,
  };
}

/**
 * Create new shipment item from order item
 */
export function createShipmentItemFromOrder(
  orderItem: { order_item_id: string; product_id: string; product_name: string; sku: string; remaining_quantity: number; unit_price: number },
  orderId: string,
  orderNumber: string
): NewShipmentItem {
  return {
    orderItemId: orderItem.order_item_id,
    orderId: orderId,
    orderNumber: orderNumber,
    productId: orderItem.product_id,
    productName: orderItem.product_name,
    sku: orderItem.sku,
    quantity: orderItem.remaining_quantity,
    maxQuantity: orderItem.remaining_quantity,
    unitPrice: orderItem.unit_price,
  };
}

File: myFinance_improved_V2/lib/features/shipment/domain/entities/shipment.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'shipment.freezed.dart';

/// Shipment status enum - matches RPC status values
enum ShipmentStatus {
  pending,
  process,
  complete,
  cancelled;

  String get label {
    switch (this) {
      case ShipmentStatus.pending:
        return 'Pending';
      case ShipmentStatus.process:
        return 'Processing';
      case ShipmentStatus.complete:
        return 'Complete';
      case ShipmentStatus.cancelled:
        return 'Cancelled';
    }
  }

  String get labelEn {
    switch (this) {
      case ShipmentStatus.pending:
        return 'Pending';
      case ShipmentStatus.process:
        return 'Processing';
      case ShipmentStatus.complete:
        return 'Complete';
      case ShipmentStatus.cancelled:
        return 'Cancelled';
    }
  }

  static ShipmentStatus fromString(String value) {
    switch (value.toLowerCase()) {
      case 'pending':
        return ShipmentStatus.pending;
      case 'process':
        return ShipmentStatus.process;
      case 'complete':
        return ShipmentStatus.complete;
      case 'cancelled':
        return ShipmentStatus.cancelled;
      default:
        return ShipmentStatus.pending;
    }
  }
}

/// Shipment Item entity
@freezed
class ShipmentItem with _$ShipmentItem {
  const ShipmentItem._();

  const factory ShipmentItem({
    required String itemId,
    required String shipmentId,
    String? orderItemId,
    String? productId,
    String? productName,
    String? sku,
    required double quantity,
    String? unit,
    double? unitPrice,
    double? totalAmount,
    String? notes,
    @Default(0) int sortOrder,
    DateTime? createdAtUtc,
  }) = _ShipmentItem;

  /// Calculate total amount if not provided
  double get calculatedTotal => totalAmount ?? (quantity * (unitPrice ?? 0));
}

/// Shipment entity (full detail)
@freezed
class Shipment with _$Shipment {
  const Shipment._();

  const factory Shipment({
    required String shipmentId,
    required String shipmentNumber,
    required String companyId,
    String? trackingNumber,
    DateTime? shippedDateUtc,
    String? supplierId,
    String? supplierName,
    Map<String, dynamic>? supplierInfo,
    @Default(ShipmentStatus.pending) ShipmentStatus status,
    String? notes,
    String? createdBy,
    DateTime? createdAtUtc,
    DateTime? updatedAtUtc,
    // Items
    @Default([]) List<ShipmentItem> items,
  }) = _Shipment;

  /// Convenience getter for id (alias for shipmentId)
  String get id => shipmentId;

  /// Check if shipment is editable
  bool get isEditable => status == ShipmentStatus.pending;

  /// Check if shipment is in progress
  bool get isInProgress => status == ShipmentStatus.process;

  /// Check if shipment is completed
  bool get isCompleted => status == ShipmentStatus.complete;

  /// Total items count
  int get itemCount => items.length;

  /// Total amount from items
  double get totalAmount =>
      items.fold(0.0, (sum, item) => sum + item.calculatedTotal);
}

/// Shipment list item (lighter version for list view) - matches RPC response
@freezed
class ShipmentListItem with _$ShipmentListItem {
  const ShipmentListItem._();

  const factory ShipmentListItem({
    required String shipmentId,
    required String shipmentNumber,
    String? trackingNumber,
    DateTime? shippedDate,
    String? supplierId,
    String? supplierName,
    required ShipmentStatus status,
    @Default(0) int itemCount,
    @Default(0) double totalAmount,
    @Default(false) bool hasOrders,
    @Default(0) int linkedOrderCount,
    String? notes,
    DateTime? createdAt,
    String? createdBy,
  }) = _ShipmentListItem;

  /// Convenience getter for id (alias for shipmentId)
  String get id => shipmentId;
}

/// Paginated response for shipment list
@freezed
class PaginatedShipmentResponse with _$PaginatedShipmentResponse {
  const factory PaginatedShipmentResponse({
    required List<ShipmentListItem> data,
    required int totalCount,
    required int limit,
    required int offset,
  }) = _PaginatedShipmentResponse;
}

// =============================================================================
// Shipment Detail V2 Entities (for inventory_get_shipment_detail_v2 RPC)
// =============================================================================

/// Supplier info for shipment detail
@freezed
class ShipmentSupplierInfo with _$ShipmentSupplierInfo {
  const factory ShipmentSupplierInfo({
    String? supplierId,
    String? supplierName,
    String? supplierPhone,
    String? supplierEmail,
    @Default(false) bool isRegisteredSupplier,
  }) = _ShipmentSupplierInfo;
}

/// Shipment detail item with variant support
@freezed
class ShipmentDetailItem with _$ShipmentDetailItem {
  const ShipmentDetailItem._();

  const factory ShipmentDetailItem({
    required String itemId,
    String? productId,
    String? variantId,
    String? productName,
    String? variantName,
    String? displayName,
    @Default(false) bool hasVariants,
    String? sku,
    @Default(0) double quantityShipped,
    @Default(0) double quantityReceived,
    @Default(0) double quantityAccepted,
    @Default(0) double quantityRejected,
    @Default(0) double quantityRemaining,
    @Default(0) double unitCost,
    @Default(0) double totalAmount,
    String? unit,
    String? notes,
    @Default(0) int sortOrder,
  }) = _ShipmentDetailItem;

  /// Get effective display name (variant name or product name)
  String get effectiveDisplayName => displayName ?? productName ?? 'Unknown Product';

  /// Check if item is fully received
  bool get isFullyReceived => quantityRemaining <= 0;

  /// Progress percentage
  double get progressPercentage =>
      quantityShipped > 0 ? (quantityReceived / quantityShipped * 100) : 0;
}

/// Receiving summary for shipment detail
@freezed
class ShipmentReceivingSummary with _$ShipmentReceivingSummary {
  const ShipmentReceivingSummary._();

  const factory ShipmentReceivingSummary({
    @Default(0) double totalShipped,
    @Default(0) double totalReceived,
    @Default(0) double totalAccepted,
    @Default(0) double totalRejected,
    @Default(0) double totalRemaining,
    @Default(0) double progressPercentage,
  }) = _ShipmentReceivingSummary;

  /// Check if fully received
  bool get isFullyReceived => totalRemaining <= 0;
}

/// Linked order info for shipment detail
@freezed
class ShipmentLinkedOrder with _$ShipmentLinkedOrder {
  const factory ShipmentLinkedOrder({
    required String orderId,
    required String orderNumber,
    String? orderDate,
    String? orderStatus,
  }) = _ShipmentLinkedOrder;
}

/// Shipment detail entity (from inventory_get_shipment_detail_v2 RPC)
@freezed
class ShipmentDetail with _$ShipmentDetail {
  const ShipmentDetail._();

  const factory ShipmentDetail({
    // Header info
    required String shipmentId,
    required String shipmentNumber,
    String? trackingNumber,
    DateTime? shippedDate,
    required ShipmentStatus status,
    String? notes,
    DateTime? createdAt,
    String? createdBy,
    DateTime? updatedAt,
    String? updatedBy,
    // Supplier info
    ShipmentSupplierInfo? supplierInfo,
    // Items
    @Default([]) List<ShipmentDetailItem> items,
    // Receiving summary
    ShipmentReceivingSummary? receivingSummary,
    // Linked orders
    @Default(false) bool hasOrders,
    @Default(0) int orderCount,
    @Default([]) List<ShipmentLinkedOrder> linkedOrders,
    // Actions
    @Default(false) bool canCancel,
  }) = _ShipmentDetail;

  /// Convenience getter for id
  String get id => shipmentId;

  /// Total items count
  int get itemCount => items.length;

  /// Total amount from items
  double get totalAmount =>
      items.fold(0.0, (sum, item) => sum + item.totalAmount);

  /// Check if shipment is editable
  bool get isEditable => status == ShipmentStatus.pending;

  /// Check if shipment is in progress
  bool get isInProgress => status == ShipmentStatus.process;

  /// Check if shipment is completed
  bool get isCompleted => status == ShipmentStatus.complete;

  /// Check if shipment is cancelled
  bool get isCancelled => status == ShipmentStatus.cancelled;

  /// Supplier name shortcut
  String? get supplierName => supplierInfo?.supplierName;

  /// Check if shipment can be cancelled (based on status or RPC flag)
  /// Returns true if canCancel flag is true OR status is pending/process
  bool get isCancellable =>
      canCancel || status == ShipmentStatus.pending || status == ShipmentStatus.process;
}

